pkg/cmd/drt-run/event.go

```diff
@@ -119,16 +119,20 @@ func (l *eventLogger) logWorkloadEvent(ev Event, workloadIdx int) {
 	ev.Source = SourceWorkload
 
 	we := &l.workloadEvents[workloadIdx]
-	we.mu.Lock()
-	idx := we.mu.startIdx
-	we.mu.startIdx = (we.mu.startIdx + 1) % perWorkloadEventRetention
-	if we.mu.numEvents != len(we.events) {
-		we.mu.numEvents++
+	var idx int
+	{
+		we.mu.Lock()
+		defer func() {
```

Comment:
nb: this defer will run at the end of the function, not at the end of the block. So the io.WriteString is also part of the mutex now.

Also you can do this, instead of the closure:

```
defer we.mu.Unlock()
```